In [1]:
import numpy as np
import pandas as pd
from sqlalchemy import create_engine
import boto3
import s3fs
import awscli
import timeit
from datetime import datetime
from pandas.api.types import is_string_dtype
from pandas.api.types import is_numeric_dtype

In [2]:
s3 = boto3.resource('s3')

for bucket in s3.buckets.all():
    print(bucket.name)

team3-final-bucket
wshih-bucket


In [3]:
my_bucket = s3.Bucket('team3-final-bucket')

In [4]:
client = boto3.client('s3')
path = 's3://team3-final-bucket/Resources_clean/'

In [5]:
objects_df = pd.read_csv(path + 'objects_clean.csv', dtype={'status':str, 'entity_id':str})
objects_df.head()

C:\Users\Fahima Farha PC\anaconda3\envs\PythonData\lib\site-packages\IPython\core\interactiveshell.py:3063: DtypeWarning: Columns (3,5,7,8,9,10,11,13,14,17,18) have mixed types.Specify dtype option on import or set low_memory=False.
  interactivity=interactivity, compiler=compiler, result=result)


Unnamed: 0 object_id entity_type parent_id                name  \
0           0       c:1     Company       NaN            Wetpaint   
1           1      c:10     Company       NaN             Flektor   
2           2     c:100     Company       NaN               There   
3           3   c:10000     Company       NaN             MYWEBBO   
4           4   c:10001     Company       NaN  THE Movie Streamer   

     category_code     status  founded_at closed_at country_code  ...  \
0              web  operating  2005-10-17       NaN          USA  ...   
1      games_video   acquired         NaN       NaN          USA  ...   
2      games_video   acquired         NaN       NaN          USA  ...   
3  network_hosting  operating  2008-07-26       NaN          NaN  ...   
4      games_video  operating  2008-07-26       NaN          NaN  ...   

  investment_rounds invested_companies first_funding_at last_funding_at  \
0                 0                  0       2005-10-01      2008-05-19   
1                 0                  0              NaN             NaN   
2                 0                  0              NaN             NaN   
3                 0                  0              NaN             NaN   
4                 0                  0              NaN             NaN   

  funding_rounds  funding_total_usd  first_milestone_at last_milestone_at  \
0              3         39750000.0          2010-09-05        2013-09-18   
1              0                0.0                 NaN               NaN   
2              0                0.0          2003-02-01        2011-09-23   
3              0                0.0                 NaN               NaN   
4              0                0.0                 NaN               NaN   

  milestones  relationships  
0          5             17  
1          0              6  
2          4             12  
3          0              0  
4          0              0  

[5 rows x 25 columns]

In [6]:
objects_df.dtypes

Unnamed: 0               int64
object_id               object
entity_type             object
parent_id               object
name                    object
category_code           object
status                  object
founded_at              object
closed_at               object
country_code            object
state_code              object
city                    object
region                  object
first_investment_at     object
last_investment_at      object
investment_rounds        int64
invested_companies       int64
first_funding_at        object
last_funding_at         object
funding_rounds           int64
funding_total_usd      float64
first_milestone_at      object
last_milestone_at       object
milestones               int64
relationships            int64
dtype: object

In [7]:
objects_df['category_code'].unique()

array(['web', 'games_video', 'network_hosting', 'advertising',
       'cleantech', nan, 'enterprise', 'other', 'consulting', 'mobile',
       'health', 'software', 'analytics', 'finance', 'education',
       'medical', 'manufacturing', 'biotech', 'ecommerce',
       'public_relations', 'hardware', 'search', 'news', 'government',
       'security', 'photo_video', 'travel', 'semiconductor', 'social',
       'legal', 'transportation', 'hospitality', 'sports', 'nonprofit',
       'fashion', 'messaging', 'music', 'automotive', 'design',
       'real_estate', 'local', 'nanotech', 'pets'], dtype=object)

In [9]:
is_string_dtype(objects_df['status'])

True

In [10]:
is_numeric_dtype(objects_df['status'])

False

In [11]:
with pd.option_context('display.max_rows', None, 'display.max_columns', None):  # more options can be specified also
    print(objects_df.dtypes)

Unnamed: 0               int64
object_id               object
entity_type             object
parent_id               object
name                    object
category_code           object
status                  object
founded_at              object
closed_at               object
country_code            object
state_code              object
city                    object
region                  object
first_investment_at     object
last_investment_at      object
investment_rounds        int64
invested_companies       int64
first_funding_at        object
last_funding_at         object
funding_rounds           int64
funding_total_usd      float64
first_milestone_at      object
last_milestone_at       object
milestones               int64
relationships            int64
dtype: object


In [12]:
objects_df.head()

Unnamed: 0 object_id entity_type parent_id                name  \
0           0       c:1     Company       NaN            Wetpaint   
1           1      c:10     Company       NaN             Flektor   
2           2     c:100     Company       NaN               There   
3           3   c:10000     Company       NaN             MYWEBBO   
4           4   c:10001     Company       NaN  THE Movie Streamer   

     category_code     status  founded_at closed_at country_code  ...  \
0              web  operating  2005-10-17       NaN          USA  ...   
1      games_video   acquired         NaN       NaN          USA  ...   
2      games_video   acquired         NaN       NaN          USA  ...   
3  network_hosting  operating  2008-07-26       NaN          NaN  ...   
4      games_video  operating  2008-07-26       NaN          NaN  ...   

  investment_rounds invested_companies first_funding_at last_funding_at  \
0                 0                  0       2005-10-01      2008-05-19   
1                 0                  0              NaN             NaN   
2                 0                  0              NaN             NaN   
3                 0                  0              NaN             NaN   
4                 0                  0              NaN             NaN   

  funding_rounds  funding_total_usd  first_milestone_at last_milestone_at  \
0              3         39750000.0          2010-09-05        2013-09-18   
1              0                0.0                 NaN               NaN   
2              0                0.0          2003-02-01        2011-09-23   
3              0                0.0                 NaN               NaN   
4              0                0.0                 NaN               NaN   

  milestones  relationships  
0          5             17  
1          0              6  
2          4             12  
3          0              0  
4          0              0  

[5 rows x 25 columns]

##### Transforming Data

In [13]:
objects_df = objects_df.rename(columns = {'id':'object_id'})
objects_df.head()

Unnamed: 0 object_id entity_type parent_id                name  \
0           0       c:1     Company       NaN            Wetpaint   
1           1      c:10     Company       NaN             Flektor   
2           2     c:100     Company       NaN               There   
3           3   c:10000     Company       NaN             MYWEBBO   
4           4   c:10001     Company       NaN  THE Movie Streamer   

     category_code     status  founded_at closed_at country_code  ...  \
0              web  operating  2005-10-17       NaN          USA  ...   
1      games_video   acquired         NaN       NaN          USA  ...   
2      games_video   acquired         NaN       NaN          USA  ...   
3  network_hosting  operating  2008-07-26       NaN          NaN  ...   
4      games_video  operating  2008-07-26       NaN          NaN  ...   

  investment_rounds invested_companies first_funding_at last_funding_at  \
0                 0                  0       2005-10-01      2008-05-19   
1                 0                  0              NaN             NaN   
2                 0                  0              NaN             NaN   
3                 0                  0              NaN             NaN   
4                 0                  0              NaN             NaN   

  funding_rounds  funding_total_usd  first_milestone_at last_milestone_at  \
0              3         39750000.0          2010-09-05        2013-09-18   
1              0                0.0                 NaN               NaN   
2              0                0.0          2003-02-01        2011-09-23   
3              0                0.0                 NaN               NaN   
4              0                0.0                 NaN               NaN   

  milestones  relationships  
0          5             17  
1          0              6  
2          4             12  
3          0              0  
4          0              0  

[5 rows x 25 columns]

##### Converting date columns into date format

In [14]:
# Converting strings to date type
objects_df['founded_at'] = pd.to_datetime(objects_df.founded_at, format='%Y-%m-%d')
objects_df['closed_at'] = pd.to_datetime(objects_df.closed_at, format='%Y-%m-%d')
objects_df['first_investment_at'] = pd.to_datetime(objects_df.first_investment_at, format='%Y-%m-%d')
objects_df['last_investment_at'] = pd.to_datetime(objects_df.last_investment_at, format='%Y-%m-%d')
objects_df['first_funding_at'] = pd.to_datetime(objects_df.first_funding_at, format='%Y-%m-%d')
objects_df['last_funding_at'] = pd.to_datetime(objects_df.last_funding_at, format='%Y-%m-%d')
objects_df['first_milestone_at'] = pd.to_datetime(objects_df.first_milestone_at, format='%Y-%m-%d')
objects_df['last_milestone_at'] = pd.to_datetime(objects_df.last_milestone_at, format='%Y-%m-%d')

In [15]:
objects_df.dtypes

Unnamed: 0                      int64
object_id                      object
entity_type                    object
parent_id                      object
name                           object
category_code                  object
status                         object
founded_at             datetime64[ns]
closed_at              datetime64[ns]
country_code                   object
state_code                     object
city                           object
region                         object
first_investment_at    datetime64[ns]
last_investment_at     datetime64[ns]
investment_rounds               int64
invested_companies              int64
first_funding_at       datetime64[ns]
last_funding_at        datetime64[ns]
funding_rounds                  int64
funding_total_usd             float64
first_milestone_at     datetime64[ns]
last_milestone_at      datetime64[ns]
milestones                      int64
relationships                   int64
dtype: object

In [16]:
# Changing columns order
objects_df.columns.to_list()

['Unnamed: 0',
 'object_id',
 'entity_type',
 'parent_id',
 'name',
 'category_code',
 'status',
 'founded_at',
 'closed_at',
 'country_code',
 'state_code',
 'city',
 'region',
 'first_investment_at',
 'last_investment_at',
 'investment_rounds',
 'invested_companies',
 'first_funding_at',
 'last_funding_at',
 'funding_rounds',
 'funding_total_usd',
 'first_milestone_at',
 'last_milestone_at',
 'milestones',
 'relationships']

In [17]:
# Adding new column for collecting start dates

objects_df["start_dates"] = ""

In [18]:
# Adding a new column for collecting end dates

objects_df["end_dates"] = ""
objects_df.head()

Unnamed: 0 object_id entity_type parent_id                name  \
0           0       c:1     Company       NaN            Wetpaint   
1           1      c:10     Company       NaN             Flektor   
2           2     c:100     Company       NaN               There   
3           3   c:10000     Company       NaN             MYWEBBO   
4           4   c:10001     Company       NaN  THE Movie Streamer   

     category_code     status founded_at closed_at country_code  ...  \
0              web  operating 2005-10-17       NaT          USA  ...   
1      games_video   acquired        NaT       NaT          USA  ...   
2      games_video   acquired        NaT       NaT          USA  ...   
3  network_hosting  operating 2008-07-26       NaT          NaN  ...   
4      games_video  operating 2008-07-26       NaT          NaN  ...   

  first_funding_at last_funding_at funding_rounds funding_total_usd  \
0       2005-10-01      2008-05-19              3        39750000.0   
1              NaT             NaT              0               0.0   
2              NaT             NaT              0               0.0   
3              NaT             NaT              0               0.0   
4              NaT             NaT              0               0.0   

  first_milestone_at  last_milestone_at  milestones relationships start_dates  \
0         2010-09-05         2013-09-18           5            17               
1                NaT                NaT           0             6               
2         2003-02-01         2011-09-23           4            12               
3                NaT                NaT           0             0               
4                NaT                NaT           0             0               

   end_dates  
0             
1             
2             
3             
4             

[5 rows x 27 columns]

In [70]:
objects_df = objects_df[['object_id',
 'entity_type',
 'parent_id',
 'name',
 'category_code',
 'status',
 'start_dates',
 'end_dates',                        
 'founded_at',
 'first_investment_at',
 'first_funding_at',
 'country_code',
 'state_code',
 'city',
 'region',
 'investment_rounds',
 'invested_companies',
 'funding_rounds',                                
 'closed_at',
 'last_investment_at',
 'last_funding_at',
 'funding_total_usd',
 'first_milestone_at',
 'last_milestone_at',
 'milestones',
 'relationships']]

In [71]:
objects_df.head()

object_id entity_type parent_id                name    category_code  \
0       c:1     Company       NaN            Wetpaint              web   
1      c:10     Company       NaN             Flektor      games_video   
2     c:100     Company       NaN               There      games_video   
3   c:10000     Company       NaN             MYWEBBO  network_hosting   
4   c:10001     Company       NaN  THE Movie Streamer      games_video   

      status start_dates end_dates founded_at first_investment_at  ...  \
0  operating                       2005-10-17                 NaT  ...   
1   acquired                              NaT                 NaT  ...   
2   acquired                              NaT                 NaT  ...   
3  operating                       2008-07-26                 NaT  ...   
4  operating                       2008-07-26                 NaT  ...   

  invested_companies funding_rounds closed_at last_investment_at  \
0                  0              3       NaT                NaT   
1                  0              0       NaT                NaT   
2                  0              0       NaT                NaT   
3                  0              0       NaT                NaT   
4                  0              0       NaT                NaT   

  last_funding_at  funding_total_usd  first_milestone_at  last_milestone_at  \
0      2008-05-19         39750000.0          2010-09-05         2013-09-18   
1             NaT                0.0                 NaT                NaT   
2             NaT                0.0          2003-02-01         2011-09-23   
3             NaT                0.0                 NaT                NaT   
4             NaT                0.0                 NaT                NaT   

  milestones relationships  
0          5            17  
1          0             6  
2          4            12  
3          0             0  
4          0             0  

[5 rows x 26 columns]

In [72]:
# Filtering 'companies' only
companies_df = objects_df[objects_df['entity_type']=='Company']
companies_df

object_id entity_type parent_id                name     category_code  \
0            c:1     Company       NaN            Wetpaint               web   
1           c:10     Company       NaN             Flektor       games_video   
2          c:100     Company       NaN               There       games_video   
3        c:10000     Company       NaN             MYWEBBO   network_hosting   
4        c:10001     Company       NaN  THE Movie Streamer       games_video   
...          ...         ...       ...                 ...               ...   
196548   c:99940     Company       NaN           Webgility         ecommerce   
196549    c:9995     Company       NaN              FohBoh  public_relations   
196550    c:9996     Company       NaN                 CSC        consulting   
196551    c:9997     Company       NaN      Top-candidate™            search   
196552    c:9998     Company       NaN      Closet Couture         ecommerce   

           status start_dates end_dates founded_at first_investment_at  ...  \
0       operating                       2005-10-17                 NaT  ...   
1        acquired                              NaT                 NaT  ...   
2        acquired                              NaT                 NaT  ...   
3       operating                       2008-07-26                 NaT  ...   
4       operating                       2008-07-26                 NaT  ...   
...           ...         ...       ...        ...                 ...  ...   
196548  operating                       2007-01-01                 NaT  ...   
196549  operating                       2007-11-01                 NaT  ...   
196550  operating                       1959-01-01                 NaT  ...   
196551  operating                       2008-07-01                 NaT  ...   
196552  operating                       2007-09-01                 NaT  ...   

       invested_companies funding_rounds closed_at last_investment_at  \
0                       0              3       NaT                NaT   
1                       0              0       NaT                NaT   
2                       0              0       NaT                NaT   
3                       0              0       NaT                NaT   
4                       0              0       NaT                NaT   
...                   ...            ...       ...                ...   
196548                  0              0       NaT                NaT   
196549                  0              1       NaT                NaT   
196550                  0              0       NaT                NaT   
196551                  0              0       NaT                NaT   
196552                  0              1       NaT                NaT   

       last_funding_at  funding_total_usd  first_milestone_at  \
0           2008-05-19         39750000.0          2010-09-05   
1                  NaT                0.0                 NaT   
2                  NaT                0.0          2003-02-01   
3                  NaT                0.0                 NaT   
4                  NaT                0.0                 NaT   
...                ...                ...                 ...   
196548             NaT                0.0          2013-08-12   
196549      2008-08-25           750000.0          2013-05-01   
196550             NaT                0.0          2012-01-01   
196551             NaT                0.0                 NaT   
196552      2008-06-01           475000.0          2006-07-01   

        last_milestone_at milestones relationships  
0              2013-09-18          5            17  
1                     NaT          0             6  
2              2011-09-23          4            12  
3                     NaT          0             0  
4                     NaT          0             0  
...                   ...        ...           ...  
196548         2013-08-12          2             5  
196549         

##### Working on filling in gap of start and end dates

In [73]:
# Filling out start dates gaps by combining dates from three columns of 'Companies':
# funded_at, first_investment_at, first_funding_at.
# If conflicting dates are found in one record, the earliest is taken.

def mindate(a,b,c):
    return min(a,b,c)

companies_df['start_dates'] = companies_df.apply(lambda row : mindate(row['founded_at'], 
                     row['first_investment_at'], row['first_funding_at']), axis = 1)

C:\Users\Fahima Farha PC\anaconda3\envs\PythonData\lib\site-packages\ipykernel_launcher.py:9: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  if __name__ == '__main__':


In [74]:
# Get the max date to fill "end_dates" gap
def maxdate(a,b,c):
    if a == 'NaT':
        return max(b,c)
    elif b == 'NaT':
        return max(a,c)
    else:
        return c
    
companies_df['end_dates'] = companies_df.apply(lambda row : maxdate(row['closed_at'], 
                      row['last_investment_at'], row['last_funding_at']), axis = 1)

C:\Users\Fahima Farha PC\anaconda3\envs\PythonData\lib\site-packages\ipykernel_launcher.py:11: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  # This is added back by InteractiveShellApp.init_path()


In [75]:
companies_df

object_id entity_type parent_id                name     category_code  \
0            c:1     Company       NaN            Wetpaint               web   
1           c:10     Company       NaN             Flektor       games_video   
2          c:100     Company       NaN               There       games_video   
3        c:10000     Company       NaN             MYWEBBO   network_hosting   
4        c:10001     Company       NaN  THE Movie Streamer       games_video   
...          ...         ...       ...                 ...               ...   
196548   c:99940     Company       NaN           Webgility         ecommerce   
196549    c:9995     Company       NaN              FohBoh  public_relations   
196550    c:9996     Company       NaN                 CSC        consulting   
196551    c:9997     Company       NaN      Top-candidate™            search   
196552    c:9998     Company       NaN      Closet Couture         ecommerce   

           status start_dates  end_dates founded_at first_investment_at  ...  \
0       operating  2005-10-01 2008-05-19 2005-10-17                 NaT  ...   
1        acquired         NaT        NaT        NaT                 NaT  ...   
2        acquired         NaT        NaT        NaT                 NaT  ...   
3       operating  2008-07-26        NaT 2008-07-26                 NaT  ...   
4       operating  2008-07-26        NaT 2008-07-26                 NaT  ...   
...           ...         ...        ...        ...                 ...  ...   
196548  operating  2007-01-01        NaT 2007-01-01                 NaT  ...   
196549  operating  2007-11-01 2008-08-25 2007-11-01                 NaT  ...   
196550  operating  1959-01-01        NaT 1959-01-01                 NaT  ...   
196551  operating  2008-07-01        NaT 2008-07-01                 NaT  ...   
196552  operating  2007-09-01 2008-06-01 2007-09-01                 NaT  ...   

       invested_companies funding_rounds closed_at last_investment_at  \
0                       0              3       NaT                NaT   
1                       0              0       NaT                NaT   
2                       0              0       NaT                NaT   
3                       0              0       NaT                NaT   
4                       0              0       NaT                NaT   
...                   ...            ...       ...                ...   
196548                  0              0       NaT                NaT   
196549                  0              1       NaT                NaT   
196550                  0              0       NaT                NaT   
196551                  0              0       NaT                NaT   
196552                  0              1       NaT                NaT   

       last_funding_at  funding_total_usd  first_milestone_at  \
0           2008-05-19         39750000.0          2010-09-05   
1                  NaT                0.0                 NaT   
2                  NaT                0.0          2003-02-01   
3                  NaT                0.0                 NaT   
4                  NaT                0.0                 NaT   
...                ...                ...                 ...   
196548             NaT                0.0          2013-08-12   
196549      2008-08-25           750000.0          2013-05-01   
196550             NaT                0.0          2012-01-01   
196551             NaT                0.0                 NaT   
196552      2008-06-01           475000.0          2006-07-01   

        last_milestone_at milestones relationships  
0              2013-09-18          5            17  
1                     NaT          0             6  
2              2011-09-23          4            12  
3                     NaT          0             0  
4                     NaT          0             0  
...                   ...        ...           ...  
196548         2013-08-12          2             5  
196

In [76]:
# Dropping columns
cleaned_companies_df = companies_df.drop(companies_df[['founded_at','first_investment_at','first_funding_at','closed_at','last_investment_at','last_funding_at']], axis=1)
cleaned_companies_df

object_id entity_type parent_id                name     category_code  \
0            c:1     Company       NaN            Wetpaint               web   
1           c:10     Company       NaN             Flektor       games_video   
2          c:100     Company       NaN               There       games_video   
3        c:10000     Company       NaN             MYWEBBO   network_hosting   
4        c:10001     Company       NaN  THE Movie Streamer       games_video   
...          ...         ...       ...                 ...               ...   
196548   c:99940     Company       NaN           Webgility         ecommerce   
196549    c:9995     Company       NaN              FohBoh  public_relations   
196550    c:9996     Company       NaN                 CSC        consulting   
196551    c:9997     Company       NaN      Top-candidate™            search   
196552    c:9998     Company       NaN      Closet Couture         ecommerce   

           status start_dates  end_dates country_code state_code  \
0       operating  2005-10-01 2008-05-19          USA         WA   
1        acquired         NaT        NaT          USA         CA   
2        acquired         NaT        NaT          USA         CA   
3       operating  2008-07-26        NaT          NaN        NaN   
4       operating  2008-07-26        NaT          NaN        NaN   
...           ...         ...        ...          ...        ...   
196548  operating  2007-01-01        NaT          USA         CA   
196549  operating  2007-11-01 2008-08-25          USA         CA   
196550  operating  1959-01-01        NaT          USA         VA   
196551  operating  2008-07-01        NaT          USA         CA   
196552  operating  2007-09-01 2008-06-01          USA         CA   

                     city             region  investment_rounds  \
0                 Seattle            Seattle                  0   
1             Culver City        Los Angeles                  0   
2               San Mateo             SF Bay                  0   
3                     NaN            unknown                  0   
4                     NaN            unknown                  0   
...                   ...                ...                ...   
196548      San Francisco             SF Bay                  0   
196549           San Jose             SF Bay                  0   
196550       Falls Church      Washington DC                  0   
196551        Los Angeles        Los Angeles                  0   
196552  Pacific Palisades  Pacific Palisades                  0   

        invested_companies  funding_rounds  funding_total_usd  \
0                        0               3         39750000.0   
1                        0               0                0.0   
2                        0               0                0.0   
3                        0               0                0.0   
4                        0               0                0.0   
...                    ...             ...                ...   
196548                   0               0                0.0   
196549                   0               1           750000.0   
196550                   0               0                0.0   
196551                   0               0                0.0   
196552                   0               1           475000.0   

       first_milestone_at last_milestone_at  milestones  relationships  
0              2010-09-05        2013-09-18           5             17  
1                     NaT               NaT           0              6  
2              2003-02-01        2011-09-23           4             12  
3                     NaT               NaT           0              0  
4                     NaT               NaT           0              0  
...                   ...               ...         ...            ...  
196548         2013-08-12        2013-08-12           2              5  
196549         2013-05-01        2013-05-01           1    

In [77]:
# Dropping nulls in end_dates
cleaned_companies_df[cleaned_companies_df['end_dates'].notna()]

object_id entity_type parent_id                           name  \
0            c:1     Company       NaN                       Wetpaint   
13        c:1001     Company       NaN                     FriendFeed   
18       c:10014     Company       NaN                        Mobclix   
19       c:10015     Company       NaN                         Fitbit   
20      c:100155     Company       NaN                           MTPV   
...          ...         ...       ...                            ...   
196537   c:99853     Company       NaN                   ChatterBlock   
196540    c:9989     Company       NaN  Shattered Reality Interactive   
196547    c:9994     Company       NaN                         Imindi   
196549    c:9995     Company       NaN                         FohBoh   
196552    c:9998     Company       NaN                 Closet Couture   

           category_code     status start_dates  end_dates country_code  \
0                    web  operating  2005-10-01 2008-05-19          USA   
13                   web   acquired  2007-10-01 2008-02-26          USA   
18                mobile   acquired  2008-03-01 2008-09-01          USA   
19                health  operating  2007-10-01 2013-08-13          USA   
20             cleantech  operating  2003-01-01 2012-01-26          USA   
...                  ...        ...         ...        ...          ...   
196537            social  operating  2011-01-11 2011-09-06          CAN   
196540       games_video     closed  2006-06-01 2009-02-01          USA   
196547       advertising  operating  2007-01-01 2008-05-01          NaN   
196549  public_relations  operating  2007-11-01 2008-08-25          USA   
196552         ecommerce  operating  2007-09-01 2008-06-01          USA   

       state_code               city             region  investment_rounds  \
0              WA            Seattle            Seattle                  0   
13             CA      Mountain View             SF Bay                  0   
18             CA          Palo Alto             SF Bay                  0   
19             CA      San Francisco             SF Bay                  0   
20             TX             Austin             Austin                  0   
...           ...                ...                ...                ...   
196537        NaN           Victoria           Victoria                  0   
196540         CA      San Francisco             SF Bay                  0   
196547        NaN                NaN            unknown                  0   
196549         CA           San Jose             SF Bay                  0   
196552         CA  Pacific Palisades  Pacific Palisades                  0   

        invested_companies  funding_rounds  funding_total_usd  \
0                        0               3         39750000.0   
13                       0               1          5000000.0   
18                       0               1                0.0   
19                       0               5         68069200.0   
20                       0               3         10125293.0   
...                    ...             ...                ...   
196537                   0               1           500000.0   
196540                   0               2          1100000.0   
196547                   0               2           500000.0   
196549                   0               1           750000.0   
196552                   0               1           475000.0   

       first_milestone_at last_milestone_at  milestones  relationships  
0              2010-09-05        2013-09-18           5             17  
13             2008-05-01        2012-09-13           3             14  
18             1995-03-01        2012-09-18           4              9  
19                    NaT               NaT           0             14  
20             2010-01-01        2010-01-01           1              6  
...                   ...               ...         ...            ...  


##### There are only 31507 end date records from the companies_df
-----

#### Looking at Product from original df for start and end dates

In [78]:
# Looking at products from the original df for end dates

end_products_df = objects_df[objects_df['entity_type']=='Product']
end_products_df

object_id entity_type parent_id  \
434913       r:1     Product       c:1   
434914      r:10     Product      c:16   
434915     r:100     Product    c:1585   
434916   r:10002     Product   c:14171   
434917   r:10003     Product   c:14173   
...          ...         ...       ...   
462646    r:9995     Product   c:14164   
462647    r:9996     Product   c:14164   
462648    r:9997     Product   c:14164   
462649    r:9998     Product   c:14164   
462650    r:9999     Product   c:14164   

                                                     name category_code  \
434913                                   Wikison Wetpaint           NaN   
434914                                 Slacker Web Player           NaN   
434915                 Vemo Enterprise Workforce Planning           NaN   
434916             dexmo.com - study and revision network           NaN   
434917                              VIrtual PBX Parachute           NaN   
...                                                   ...           ...   
462646              SiteLink, listing feed for Brokerages           NaN   
462647  EDCLink, listing feed for Economic Development...           NaN   
462648                   Cmail, broadcast email marketing           NaN   
462649                      CatylistCRM, contact database           NaN   
462650                   Catylist Custom Print/Web Design           NaN   

           status start_dates end_dates founded_at first_investment_at  ...  \
434913       live                       2006-06-01                 NaT  ...   
434914  operating                       2007-06-05                 NaT  ...   
434915       live                       2008-07-01                 NaT  ...   
434916  operating                              NaT                 NaT  ...   
434917  operating                              NaT                 NaT  ...   
...           ...         ...       ...        ...                 ...  ...   
462646  operating                              NaT                 NaT  ...   
462647  operating                              NaT                 NaT  ...   
462648  operating                              NaT                 NaT  ...   
462649  operating                              NaT                 NaT  ...   
462650  operating                              NaT                 NaT  ...   

       invested_companies funding_rounds closed_at last_investment_at  \
434913                  0              0       NaT                NaT   
434914                  0              0       NaT                NaT   
434915                  0              0       NaT                NaT   
434916                  0              0       NaT                NaT   
434917                  0              0       NaT                NaT   
...                   ...            ...       ...                ...   
462646                  0              0       NaT                NaT   
462647                  0              0       NaT                NaT   
462648                  0              0       NaT                NaT   
462649                  0              0       NaT                NaT   
462650                  0              0       NaT                NaT   

       last_funding_at  funding_total_usd  first_milestone_at  \
434913             NaT                0.0                 NaT   
434914             NaT                0.0                 NaT   
434915             NaT                0.0                 NaT   
434916             NaT                0.0                 NaT   
434917             NaT                0.0                 NaT   
...                ...                ...                 ...   
462646             NaT                0.0                 NaT   
462647             NaT                0.0                 NaT   
462648             NaT                0.0                 NaT   
462649             NaT                0.0                 NaT   
462650             NaT                0.0                 Na

In [79]:
# Dropping null values in parent_id
cleaned_end_products_df = end_products_df[end_products_df['parent_id'].notnull()]
cleaned_end_products_df

object_id entity_type parent_id  \
434913       r:1     Product       c:1   
434914      r:10     Product      c:16   
434915     r:100     Product    c:1585   
434916   r:10002     Product   c:14171   
434917   r:10003     Product   c:14173   
...          ...         ...       ...   
462646    r:9995     Product   c:14164   
462647    r:9996     Product   c:14164   
462648    r:9997     Product   c:14164   
462649    r:9998     Product   c:14164   
462650    r:9999     Product   c:14164   

                                                     name category_code  \
434913                                   Wikison Wetpaint           NaN   
434914                                 Slacker Web Player           NaN   
434915                 Vemo Enterprise Workforce Planning           NaN   
434916             dexmo.com - study and revision network           NaN   
434917                              VIrtual PBX Parachute           NaN   
...                                                   ...           ...   
462646              SiteLink, listing feed for Brokerages           NaN   
462647  EDCLink, listing feed for Economic Development...           NaN   
462648                   Cmail, broadcast email marketing           NaN   
462649                      CatylistCRM, contact database           NaN   
462650                   Catylist Custom Print/Web Design           NaN   

           status start_dates end_dates founded_at first_investment_at  ...  \
434913       live                       2006-06-01                 NaT  ...   
434914  operating                       2007-06-05                 NaT  ...   
434915       live                       2008-07-01                 NaT  ...   
434916  operating                              NaT                 NaT  ...   
434917  operating                              NaT                 NaT  ...   
...           ...         ...       ...        ...                 ...  ...   
462646  operating                              NaT                 NaT  ...   
462647  operating                              NaT                 NaT  ...   
462648  operating                              NaT                 NaT  ...   
462649  operating                              NaT                 NaT  ...   
462650  operating                              NaT                 NaT  ...   

       invested_companies funding_rounds closed_at last_investment_at  \
434913                  0              0       NaT                NaT   
434914                  0              0       NaT                NaT   
434915                  0              0       NaT                NaT   
434916                  0              0       NaT                NaT   
434917                  0              0       NaT                NaT   
...                   ...            ...       ...                ...   
462646                  0              0       NaT                NaT   
462647                  0              0       NaT                NaT   
462648                  0              0       NaT                NaT   
462649                  0              0       NaT                NaT   
462650                  0              0       NaT                NaT   

       last_funding_at  funding_total_usd  first_milestone_at  \
434913             NaT                0.0                 NaT   
434914             NaT                0.0                 NaT   
434915             NaT                0.0                 NaT   
434916             NaT                0.0                 NaT   
434917             NaT                0.0                 NaT   
...                ...                ...                 ...   
462646             NaT                0.0                 NaT   
462647             NaT                0.0                 NaT   
462648             NaT                0.0                 NaT   
462649             NaT                0.0                 NaT   
462650             NaT                0.0                 Na

In [81]:
cleaned_end_products_df.columns.to_list

<bound method IndexOpsMixin.tolist of Index(['object_id', 'entity_type', 'parent_id', 'name', 'category_code',
       'status', 'start_dates', 'end_dates', 'founded_at',
       'first_investment_at', 'first_funding_at', 'country_code', 'state_code',
       'city', 'region', 'investment_rounds', 'invested_companies',
       'funding_rounds', 'closed_at', 'last_investment_at', 'last_funding_at',
       'funding_total_usd', 'first_milestone_at', 'last_milestone_at',
       'milestones', 'relationships'],
      dtype='object')>

In [82]:
cleaned_end_products_df['founded_at'].unique()

array(['2006-06-01T00:00:00.000000000', '2007-06-05T00:00:00.000000000',
       '2008-07-01T00:00:00.000000000', ...,
       '2007-03-15T00:00:00.000000000', '2008-11-27T00:00:00.000000000',
       '2007-12-30T00:00:00.000000000'], dtype='datetime64[ns]')

In [83]:
cleaned_end_products_df['first_investment_at'].unique()

array(['NaT'], dtype='datetime64[ns]')

In [84]:
cleaned_end_products_df['first_funding_at'].unique()

array(['NaT'], dtype='datetime64[ns]')

In [85]:
# Dropping all redundant columns

ce2_products_df = cleaned_end_products_df.drop(['entity_type',  
       'category_code', 'first_investment_at', 'first_funding_at', 'country_code', 'state_code',
       'city', 'region', 'investment_rounds', 'invested_companies',
       'funding_rounds', 'funding_total_usd', 'first_milestone_at', 'last_milestone_at',
       'milestones', 'relationships'], axis=1)

ce2_products_df

object_id parent_id                                               name  \
434913       r:1       c:1                                   Wikison Wetpaint   
434914      r:10      c:16                                 Slacker Web Player   
434915     r:100    c:1585                 Vemo Enterprise Workforce Planning   
434916   r:10002   c:14171             dexmo.com - study and revision network   
434917   r:10003   c:14173                              VIrtual PBX Parachute   
...          ...       ...                                                ...   
462646    r:9995   c:14164              SiteLink, listing feed for Brokerages   
462647    r:9996   c:14164  EDCLink, listing feed for Economic Development...   
462648    r:9997   c:14164                   Cmail, broadcast email marketing   
462649    r:9998   c:14164                      CatylistCRM, contact database   
462650    r:9999   c:14164                   Catylist Custom Print/Web Design   

           status start_dates end_dates founded_at closed_at  \
434913       live                       2006-06-01       NaT   
434914  operating                       2007-06-05       NaT   
434915       live                       2008-07-01       NaT   
434916  operating                              NaT       NaT   
434917  operating                              NaT       NaT   
...           ...         ...       ...        ...       ...   
462646  operating                              NaT       NaT   
462647  operating                              NaT       NaT   
462648  operating                              NaT       NaT   
462649  operating                              NaT       NaT   
462650  operating                              NaT       NaT   

       last_investment_at last_funding_at  
434913                NaT             NaT  
434914                NaT             NaT  
434915                NaT             NaT  
434916                NaT             NaT  
434917                NaT             NaT  
...                   ...             ...  
462646                NaT             NaT  
462647                NaT             NaT  
462648                NaT             NaT  
462649                NaT             NaT  
462650                NaT             NaT  

[27715 rows x 10 columns]

In [86]:
ce2_products_df['last_investment_at'].unique()

array(['NaT'], dtype='datetime64[ns]')

In [87]:
ce2_products_df['last_funding_at'].unique()

array(['NaT'], dtype='datetime64[ns]')

In [88]:
# Dropping last_investment_at and last_funding_at columns

products_df_final = ce2_products_df.drop(['last_investment_at', 'last_funding_at'], axis=1)
products_df_final

object_id parent_id                                               name  \
434913       r:1       c:1                                   Wikison Wetpaint   
434914      r:10      c:16                                 Slacker Web Player   
434915     r:100    c:1585                 Vemo Enterprise Workforce Planning   
434916   r:10002   c:14171             dexmo.com - study and revision network   
434917   r:10003   c:14173                              VIrtual PBX Parachute   
...          ...       ...                                                ...   
462646    r:9995   c:14164              SiteLink, listing feed for Brokerages   
462647    r:9996   c:14164  EDCLink, listing feed for Economic Development...   
462648    r:9997   c:14164                   Cmail, broadcast email marketing   
462649    r:9998   c:14164                      CatylistCRM, contact database   
462650    r:9999   c:14164                   Catylist Custom Print/Web Design   

           status start_dates end_dates founded_at closed_at  
434913       live                       2006-06-01       NaT  
434914  operating                       2007-06-05       NaT  
434915       live                       2008-07-01       NaT  
434916  operating                              NaT       NaT  
434917  operating                              NaT       NaT  
...           ...         ...       ...        ...       ...  
462646  operating                              NaT       NaT  
462647  operating                              NaT       NaT  
462648  operating                              NaT       NaT  
462649  operating                              NaT       NaT  
462650  operating                              NaT       NaT  

[27715 rows x 8 columns]

In [89]:
products_df_final = products_df_final[products_df_final['closed_at'].notna()]
products_df_final

object_id parent_id               name  status start_dates end_dates  \
435039   r:10173   c:14409        Friendzbook  closed                         
435064     r:102     c:460       Sticky Notes  closed                         
435118   r:10253    c:3347           ZuiGames  closed                         
435229   r:10393   c:14624            Bitloot  closed                         
435236     r:104     c:112               Zlio  closed                         
...          ...       ...                ...     ...         ...       ...   
461825    r:8963   c:11424        PlayCrafter  closed                         
461911    r:9055   c:12152  MacFeeder Sverige  closed                         
462096    r:9291      c:12             Twicco  closed                         
462124    r:9320   c:13193    WW.DataServices  closed                         
462194    r:9404   c:13337            Artemis  closed                         

       founded_at  closed_at  
435039 2007-09-21 2007-09-21  
435064 2007-08-18 2010-01-01  
435118 2008-11-14 2008-11-14  
435229 2008-11-14 2009-04-01  
435236 2006-09-10 2011-09-20  
...           ...        ...  
461825 2008-07-22 1993-05-12  
461911 2008-07-01 2009-06-01  
462096 2009-11-04 2009-07-31  
462124 2006-04-01 2007-03-01  
462194 2004-01-01 2005-01-01  

[189 rows x 8 columns]

In [90]:
products_df_final_2 = products_df_final[products_df_final['founded_at'].notna()]
products_df_final_2

object_id parent_id               name  status start_dates end_dates  \
435039   r:10173   c:14409        Friendzbook  closed                         
435064     r:102     c:460       Sticky Notes  closed                         
435118   r:10253    c:3347           ZuiGames  closed                         
435229   r:10393   c:14624            Bitloot  closed                         
435236     r:104     c:112               Zlio  closed                         
...          ...       ...                ...     ...         ...       ...   
461825    r:8963   c:11424        PlayCrafter  closed                         
461911    r:9055   c:12152  MacFeeder Sverige  closed                         
462096    r:9291      c:12             Twicco  closed                         
462124    r:9320   c:13193    WW.DataServices  closed                         
462194    r:9404   c:13337            Artemis  closed                         

       founded_at  closed_at  
435039 2007-09-21 2007-09-21  
435064 2007-08-18 2010-01-01  
435118 2008-11-14 2008-11-14  
435229 2008-11-14 2009-04-01  
435236 2006-09-10 2011-09-20  
...           ...        ...  
461825 2008-07-22 1993-05-12  
461911 2008-07-01 2009-06-01  
462096 2009-11-04 2009-07-31  
462124 2006-04-01 2007-03-01  
462194 2004-01-01 2005-01-01  

[151 rows x 8 columns]

##### Only 151 more start-end_dates pairs would be available from the products_df

In [91]:
# Final product df with start and end dates
products_df_final_2['end_dates'] = products_df_final_2.apply(lambda row : (row['closed_at']), axis = 1)
products_df_final_2

C:\Users\Fahima Farha PC\anaconda3\envs\PythonData\lib\site-packages\ipykernel_launcher.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  


object_id parent_id               name  status start_dates  end_dates  \
435039   r:10173   c:14409        Friendzbook  closed             2007-09-21   
435064     r:102     c:460       Sticky Notes  closed             2010-01-01   
435118   r:10253    c:3347           ZuiGames  closed             2008-11-14   
435229   r:10393   c:14624            Bitloot  closed             2009-04-01   
435236     r:104     c:112               Zlio  closed             2011-09-20   
...          ...       ...                ...     ...         ...        ...   
461825    r:8963   c:11424        PlayCrafter  closed             1993-05-12   
461911    r:9055   c:12152  MacFeeder Sverige  closed             2009-06-01   
462096    r:9291      c:12             Twicco  closed             2009-07-31   
462124    r:9320   c:13193    WW.DataServices  closed             2007-03-01   
462194    r:9404   c:13337            Artemis  closed             2005-01-01   

       founded_at  closed_at  
435039 2007-09-21 2007-09-21  
435064 2007-08-18 2010-01-01  
435118 2008-11-14 2008-11-14  
435229 2008-11-14 2009-04-01  
435236 2006-09-10 2011-09-20  
...           ...        ...  
461825 2008-07-22 1993-05-12  
461911 2008-07-01 2009-06-01  
462096 2009-11-04 2009-07-31  
462124 2006-04-01 2007-03-01  
462194 2004-01-01 2005-01-01  

[151 rows x 8 columns]

In [93]:
# Final product df with start and end dates
products_df_final_2['start_dates'] = products_df_final_2.apply(lambda row : (row['founded_at']), axis = 1)
products_df_final_2['end_dates'] = products_df_final_2.apply(lambda row : (row['closed_at']), axis = 1)
products_df_final_2

C:\Users\Fahima Farha PC\anaconda3\envs\PythonData\lib\site-packages\ipykernel_launcher.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  
C:\Users\Fahima Farha PC\anaconda3\envs\PythonData\lib\site-packages\ipykernel_launcher.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  This is separate from the ipykernel package so we can avoid doing imports until


object_id parent_id               name  status start_dates  end_dates  \
435039   r:10173   c:14409        Friendzbook  closed  2007-09-21 2007-09-21   
435064     r:102     c:460       Sticky Notes  closed  2007-08-18 2010-01-01   
435118   r:10253    c:3347           ZuiGames  closed  2008-11-14 2008-11-14   
435229   r:10393   c:14624            Bitloot  closed  2008-11-14 2009-04-01   
435236     r:104     c:112               Zlio  closed  2006-09-10 2011-09-20   
...          ...       ...                ...     ...         ...        ...   
461825    r:8963   c:11424        PlayCrafter  closed  2008-07-22 1993-05-12   
461911    r:9055   c:12152  MacFeeder Sverige  closed  2008-07-01 2009-06-01   
462096    r:9291      c:12             Twicco  closed  2009-11-04 2009-07-31   
462124    r:9320   c:13193    WW.DataServices  closed  2006-04-01 2007-03-01   
462194    r:9404   c:13337            Artemis  closed  2004-01-01 2005-01-01   

       founded_at  closed_at  
435039 2007-09-21 2007-09-21  
435064 2007-08-18 2010-01-01  
435118 2008-11-14 2008-11-14  
435229 2008-11-14 2009-04-01  
435236 2006-09-10 2011-09-20  
...           ...        ...  
461825 2008-07-22 1993-05-12  
461911 2008-07-01 2009-06-01  
462096 2009-11-04 2009-07-31  
462124 2006-04-01 2007-03-01  
462194 2004-01-01 2005-01-01  

[151 rows x 8 columns]

In [94]:
#Dropping columns
fp_df = products_df_final_2.drop(products_df_final_2[['closed_at', 'founded_at']], axis=1)
fp_df

object_id parent_id               name  status start_dates  end_dates
435039   r:10173   c:14409        Friendzbook  closed  2007-09-21 2007-09-21
435064     r:102     c:460       Sticky Notes  closed  2007-08-18 2010-01-01
435118   r:10253    c:3347           ZuiGames  closed  2008-11-14 2008-11-14
435229   r:10393   c:14624            Bitloot  closed  2008-11-14 2009-04-01
435236     r:104     c:112               Zlio  closed  2006-09-10 2011-09-20
...          ...       ...                ...     ...         ...        ...
461825    r:8963   c:11424        PlayCrafter  closed  2008-07-22 1993-05-12
461911    r:9055   c:12152  MacFeeder Sverige  closed  2008-07-01 2009-06-01
462096    r:9291      c:12             Twicco  closed  2009-11-04 2009-07-31
462124    r:9320   c:13193    WW.DataServices  closed  2006-04-01 2007-03-01
462194    r:9404   c:13337            Artemis  closed  2004-01-01 2005-01-01

[151 rows x 6 columns]

In [95]:
merged_df = pd.merge(cleaned_companies_df, fp_df, how='left', on=['object_id'])
merged_df

object_id entity_type parent_id_x              name_x  \
0            c:1     Company         NaN            Wetpaint   
1           c:10     Company         NaN             Flektor   
2          c:100     Company         NaN               There   
3        c:10000     Company         NaN             MYWEBBO   
4        c:10001     Company         NaN  THE Movie Streamer   
...          ...         ...         ...                 ...   
196548   c:99940     Company         NaN           Webgility   
196549    c:9995     Company         NaN              FohBoh   
196550    c:9996     Company         NaN                 CSC   
196551    c:9997     Company         NaN      Top-candidate™   
196552    c:9998     Company         NaN      Closet Couture   

           category_code   status_x start_dates_x end_dates_x country_code  \
0                    web  operating    2005-10-01  2008-05-19          USA   
1            games_video   acquired           NaT         NaT          USA   
2            games_video   acquired           NaT         NaT          USA   
3        network_hosting  operating    2008-07-26         NaT          NaN   
4            games_video  operating    2008-07-26         NaT          NaN   
...                  ...        ...           ...         ...          ...   
196548         ecommerce  operating    2007-01-01         NaT          USA   
196549  public_relations  operating    2007-11-01  2008-08-25          USA   
196550        consulting  operating    1959-01-01         NaT          USA   
196551            search  operating    2008-07-01         NaT          USA   
196552         ecommerce  operating    2007-09-01  2008-06-01          USA   

       state_code  ... funding_total_usd first_milestone_at  \
0              WA  ...        39750000.0         2010-09-05   
1              CA  ...               0.0                NaT   
2              CA  ...               0.0         2003-02-01   
3             NaN  ...               0.0                NaT   
4             NaN  ...               0.0                NaT   
...           ...  ...               ...                ...   
196548         CA  ...               0.0         2013-08-12   
196549         CA  ...          750000.0         2013-05-01   
196550         VA  ...               0.0         2012-01-01   
196551         CA  ...               0.0                NaT   
196552         CA  ...          475000.0         2006-07-01   

        last_milestone_at  milestones  relationships  parent_id_y name_y  \
0              2013-09-18           5             17          NaN    NaN   
1                     NaT           0              6          NaN    NaN   
2              2011-09-23           4             12          NaN    NaN   
3                     NaT           0              0          NaN    NaN   
4                     NaT           0              0          NaN    NaN   
...                   ...         ...            ...          ...    ...   
196548         2013-08-12           2              5          NaN    NaN   
196549         2013-05-01           1             14          NaN    NaN   
196550         2013-10-30           3             44          NaN    NaN   
196551                NaT           0              1          NaN    NaN   
196552         2008-09-10           2              1          NaN    NaN   

       status_y  start_dates_y  end_dates_y  
0           NaN            NaT          NaT  
1           NaN            NaT          NaT  
2           NaN            NaT          NaT  
3           NaN            NaT          NaT  
4           NaN            NaT          NaT  
...         ...            ...          ...  
196548      NaN            NaT          NaT  
196549      NaN            NaT          NaT  
196550      NaN            NaT          NaT  
196551      NaN            NaT          NaT  
196552      NaN            NaT          NaT  

[196553 rows x 25 columns]

In [96]:
merged_df.columns.to_list

<bound method IndexOpsMixin.tolist of Index(['object_id', 'entity_type', 'parent_id_x', 'name_x', 'category_code',
       'status_x', 'start_dates_x', 'end_dates_x', 'country_code',
       'state_code', 'city', 'region', 'investment_rounds',
       'invested_companies', 'funding_rounds', 'funding_total_usd',
       'first_milestone_at', 'last_milestone_at', 'milestones',
       'relationships', 'parent_id_y', 'name_y', 'status_y', 'start_dates_y',
       'end_dates_y'],
      dtype='object')>

In [97]:
# checking to see if there are any not NaT values
merged_df['start_dates_y'].unique()

array(['NaT'], dtype='datetime64[ns]')

In [98]:
# checking for any unique values
merged_df['end_dates_y'].unique()

array(['NaT'], dtype='datetime64[ns]')

In [99]:
# Dropping all redundant columns

clean_merge_df = merged_df.drop(['parent_id_y',
       'name_y', 'status_y', 'start_dates_y', 'end_dates_y'], axis=1)

clean_merge_df

object_id entity_type parent_id_x              name_x  \
0            c:1     Company         NaN            Wetpaint   
1           c:10     Company         NaN             Flektor   
2          c:100     Company         NaN               There   
3        c:10000     Company         NaN             MYWEBBO   
4        c:10001     Company         NaN  THE Movie Streamer   
...          ...         ...         ...                 ...   
196548   c:99940     Company         NaN           Webgility   
196549    c:9995     Company         NaN              FohBoh   
196550    c:9996     Company         NaN                 CSC   
196551    c:9997     Company         NaN      Top-candidate™   
196552    c:9998     Company         NaN      Closet Couture   

           category_code   status_x start_dates_x end_dates_x country_code  \
0                    web  operating    2005-10-01  2008-05-19          USA   
1            games_video   acquired           NaT         NaT          USA   
2            games_video   acquired           NaT         NaT          USA   
3        network_hosting  operating    2008-07-26         NaT          NaN   
4            games_video  operating    2008-07-26         NaT          NaN   
...                  ...        ...           ...         ...          ...   
196548         ecommerce  operating    2007-01-01         NaT          USA   
196549  public_relations  operating    2007-11-01  2008-08-25          USA   
196550        consulting  operating    1959-01-01         NaT          USA   
196551            search  operating    2008-07-01         NaT          USA   
196552         ecommerce  operating    2007-09-01  2008-06-01          USA   

       state_code               city             region  investment_rounds  \
0              WA            Seattle            Seattle                  0   
1              CA        Culver City        Los Angeles                  0   
2              CA          San Mateo             SF Bay                  0   
3             NaN                NaN            unknown                  0   
4             NaN                NaN            unknown                  0   
...           ...                ...                ...                ...   
196548         CA      San Francisco             SF Bay                  0   
196549         CA           San Jose             SF Bay                  0   
196550         VA       Falls Church      Washington DC                  0   
196551         CA        Los Angeles        Los Angeles                  0   
196552         CA  Pacific Palisades  Pacific Palisades                  0   

        invested_companies  funding_rounds  funding_total_usd  \
0                        0               3         39750000.0   
1                        0               0                0.0   
2                        0               0                0.0   
3                        0               0                0.0   
4                        0               0                0.0   
...                    ...             ...                ...   
196548                   0               0                0.0   
196549                   0               1           750000.0   
196550                   0               0                0.0   
196551                   0               0                0.0   
196552                   0               1           475000.0   

       first_milestone_at last_milestone_at  milestones  relationships  
0              2010-09-05        2013-09-18           5             17  
1                     NaT               NaT           0              6  
2              2003-02-01        2011-09-23           4             12  
3                     NaT               NaT           0              0  
4                     NaT               NaT           0              0  
...                   ...               ...         ...            ...  
196548         2013-08-12        2013-08-12           2              5  

In [100]:
clean_merge_df = clean_merge_df.rename(columns = {'parent_id_x': 'parent_id', 
                'name_x':'name', 'status_x':'status', 'start_dates_x':'start_dates', 'end_dates_x':'end_dates'})

clean_merge_df

object_id entity_type parent_id                name     category_code  \
0            c:1     Company       NaN            Wetpaint               web   
1           c:10     Company       NaN             Flektor       games_video   
2          c:100     Company       NaN               There       games_video   
3        c:10000     Company       NaN             MYWEBBO   network_hosting   
4        c:10001     Company       NaN  THE Movie Streamer       games_video   
...          ...         ...       ...                 ...               ...   
196548   c:99940     Company       NaN           Webgility         ecommerce   
196549    c:9995     Company       NaN              FohBoh  public_relations   
196550    c:9996     Company       NaN                 CSC        consulting   
196551    c:9997     Company       NaN      Top-candidate™            search   
196552    c:9998     Company       NaN      Closet Couture         ecommerce   

           status start_dates  end_dates country_code state_code  \
0       operating  2005-10-01 2008-05-19          USA         WA   
1        acquired         NaT        NaT          USA         CA   
2        acquired         NaT        NaT          USA         CA   
3       operating  2008-07-26        NaT          NaN        NaN   
4       operating  2008-07-26        NaT          NaN        NaN   
...           ...         ...        ...          ...        ...   
196548  operating  2007-01-01        NaT          USA         CA   
196549  operating  2007-11-01 2008-08-25          USA         CA   
196550  operating  1959-01-01        NaT          USA         VA   
196551  operating  2008-07-01        NaT          USA         CA   
196552  operating  2007-09-01 2008-06-01          USA         CA   

                     city             region  investment_rounds  \
0                 Seattle            Seattle                  0   
1             Culver City        Los Angeles                  0   
2               San Mateo             SF Bay                  0   
3                     NaN            unknown                  0   
4                     NaN            unknown                  0   
...                   ...                ...                ...   
196548      San Francisco             SF Bay                  0   
196549           San Jose             SF Bay                  0   
196550       Falls Church      Washington DC                  0   
196551        Los Angeles        Los Angeles                  0   
196552  Pacific Palisades  Pacific Palisades                  0   

        invested_companies  funding_rounds  funding_total_usd  \
0                        0               3         39750000.0   
1                        0               0                0.0   
2                        0               0                0.0   
3                        0               0                0.0   
4                        0               0                0.0   
...                    ...             ...                ...   
196548                   0               0                0.0   
196549                   0               1           750000.0   
196550                   0               0                0.0   
196551                   0               0                0.0   
196552                   0               1           475000.0   

       first_milestone_at last_milestone_at  milestones  relationships  
0              2010-09-05        2013-09-18           5             17  
1                     NaT               NaT           0              6  
2              2003-02-01        2011-09-23           4             12  
3                     NaT               NaT           0              0  
4                     NaT               NaT           0              0  
...                   ...               ...         ...            ...  
196548         2013-08-12        2013-08-12           2              5  
196549         2013-05-01        2013-05-01           1    

In [101]:
clean_merge_df.dtypes

object_id                     object
entity_type                   object
parent_id                     object
name                          object
category_code                 object
status                        object
start_dates           datetime64[ns]
end_dates             datetime64[ns]
country_code                  object
state_code                    object
city                          object
region                        object
investment_rounds              int64
invested_companies             int64
funding_rounds                 int64
funding_total_usd            float64
first_milestone_at    datetime64[ns]
last_milestone_at     datetime64[ns]
milestones                     int64
relationships                  int64
dtype: object

In [102]:
clean_merge_df

object_id entity_type parent_id                name     category_code  \
0            c:1     Company       NaN            Wetpaint               web   
1           c:10     Company       NaN             Flektor       games_video   
2          c:100     Company       NaN               There       games_video   
3        c:10000     Company       NaN             MYWEBBO   network_hosting   
4        c:10001     Company       NaN  THE Movie Streamer       games_video   
...          ...         ...       ...                 ...               ...   
196548   c:99940     Company       NaN           Webgility         ecommerce   
196549    c:9995     Company       NaN              FohBoh  public_relations   
196550    c:9996     Company       NaN                 CSC        consulting   
196551    c:9997     Company       NaN      Top-candidate™            search   
196552    c:9998     Company       NaN      Closet Couture         ecommerce   

           status start_dates  end_dates country_code state_code  \
0       operating  2005-10-01 2008-05-19          USA         WA   
1        acquired         NaT        NaT          USA         CA   
2        acquired         NaT        NaT          USA         CA   
3       operating  2008-07-26        NaT          NaN        NaN   
4       operating  2008-07-26        NaT          NaN        NaN   
...           ...         ...        ...          ...        ...   
196548  operating  2007-01-01        NaT          USA         CA   
196549  operating  2007-11-01 2008-08-25          USA         CA   
196550  operating  1959-01-01        NaT          USA         VA   
196551  operating  2008-07-01        NaT          USA         CA   
196552  operating  2007-09-01 2008-06-01          USA         CA   

                     city             region  investment_rounds  \
0                 Seattle            Seattle                  0   
1             Culver City        Los Angeles                  0   
2               San Mateo             SF Bay                  0   
3                     NaN            unknown                  0   
4                     NaN            unknown                  0   
...                   ...                ...                ...   
196548      San Francisco             SF Bay                  0   
196549           San Jose             SF Bay                  0   
196550       Falls Church      Washington DC                  0   
196551        Los Angeles        Los Angeles                  0   
196552  Pacific Palisades  Pacific Palisades                  0   

        invested_companies  funding_rounds  funding_total_usd  \
0                        0               3         39750000.0   
1                        0               0                0.0   
2                        0               0                0.0   
3                        0               0                0.0   
4                        0               0                0.0   
...                    ...             ...                ...   
196548                   0               0                0.0   
196549                   0               1           750000.0   
196550                   0               0                0.0   
196551                   0               0                0.0   
196552                   0               1           475000.0   

       first_milestone_at last_milestone_at  milestones  relationships  
0              2010-09-05        2013-09-18           5             17  
1                     NaT               NaT           0              6  
2              2003-02-01        2011-09-23           4             12  
3                     NaT               NaT           0              0  
4                     NaT               NaT           0              0  
...                   ...               ...         ...            ...  
196548         2013-08-12        2013-08-12           2              5  
196549         2013-05-01        2013-05-01           1    

In [220]:
# Do Not run
#final_df = pd.DataFrame(list(filter(lambda x: str(x[6]) != 'NaT' and str(x[14]) != 'NaT', clean_merge_df.values.tolist())))

In [221]:
# Do Not run
#final_df

0        1   2                              3                 4  \
0           c:1  Company NaN                       Wetpaint               web   
1        c:1001  Company NaN                     FriendFeed               web   
2       c:10014  Company NaN                        Mobclix            mobile   
3       c:10015  Company NaN                         Fitbit            health   
4      c:100155  Company NaN                           MTPV         cleantech   
...         ...      ...  ..                            ...               ...   
24296   c:99853  Company NaN                   ChatterBlock            social   
24297    c:9989  Company NaN  Shattered Reality Interactive       games_video   
24298    c:9994  Company NaN                         Imindi       advertising   
24299    c:9995  Company NaN                         FohBoh  public_relations   
24300    c:9998  Company NaN                 Closet Couture         ecommerce   

               5          6    7    8        9                 10  11  12  13  \
0      operating 2005-10-01  USA   WA  Seattle            Seattle   0   0   3   
1       acquired 2007-10-01  USA   CA  Seattle             SF Bay   0   0   1   
2       acquired 2008-03-01  USA   CA  Seattle             SF Bay   0   0   1   
3      operating 2007-10-01  USA   CA  Seattle             SF Bay   0   0   5   
4      operating 2003-01-01  USA   TX  Seattle             Austin   0   0   3   
...          ...        ...  ...  ...      ...                ...  ..  ..  ..   
24296  operating 2011-01-11  CAN  NaN  Seattle           Victoria   0   0   1   
24297     closed 2006-06-01  USA   CA  Seattle             SF Bay   0   0   2   
24298  operating 2007-01-01  NaN  NaN  Seattle            unknown   0   0   2   
24299  operating 2007-11-01  USA   CA  Seattle             SF Bay   0   0   1   
24300  operating 2007-09-01  USA   CA  Seattle  Pacific Palisades   0   0   1   

              14          15         16         17  18  19  
0     2008-05-19  39750000.0 2010-09-05 2013-09-18   5  17  
1     2008-02-26   5000000.0 2008-05-01 2012-09-13   3  14  
2     2008-09-01         0.0 1995-03-01 2012-09-18   4   9  
3     2013-08-13  68069200.0        NaT        NaT   0  14  
4     2012-01-26  10125293.0 2010-01-01 2010-01-01   1   6  
...          ...         ...        ...        ...  ..  ..  
24296 2011-09-06    500000.0 2011-01-11 2011-01-11   1   2  
24297 2009-02-01   1100000.0 2005-11-01 2008-09-10   2   4  
24298 2008-05-01    500000.0 2009-08-27 2009-08-27   1   2  
24299 2008-08-25    750000.0 2013-05-01 2013-05-01   1  14  
24300 2008-06-01    475000.0 2006-07-01 2008-09-10   2   1  

[24301 rows x 20 columns]

Convert column names into df then use it in list conversion

In [182]:
# final_df = clean_merge_df[(clean_merge_df['start_dates'] != '<M8[NaT]') and (clean_merge_df['end_dates'] != '<M8[NaT]')



# final_df = clean_merge_df[!pd.isnull(np.datetime64(clean_merge_df['start_dates'])]
# final_df    

ValueError: The truth value of a Series is ambiguous. Use a.empty, a.bool(), a.item(), a.any() or a.all().

In [103]:
clean_merge_df_test = clean_merge_df[clean_merge_df['start_dates'].notna() & clean_merge_df['end_dates'].notna()] 
clean_merge_df_test

object_id entity_type parent_id                           name  \
0            c:1     Company       NaN                       Wetpaint   
13        c:1001     Company       NaN                     FriendFeed   
18       c:10014     Company       NaN                        Mobclix   
19       c:10015     Company       NaN                         Fitbit   
20      c:100155     Company       NaN                           MTPV   
...          ...         ...       ...                            ...   
196537   c:99853     Company       NaN                   ChatterBlock   
196540    c:9989     Company       NaN  Shattered Reality Interactive   
196547    c:9994     Company       NaN                         Imindi   
196549    c:9995     Company       NaN                         FohBoh   
196552    c:9998     Company       NaN                 Closet Couture   

           category_code     status start_dates  end_dates country_code  \
0                    web  operating  2005-10-01 2008-05-19          USA   
13                   web   acquired  2007-10-01 2008-02-26          USA   
18                mobile   acquired  2008-03-01 2008-09-01          USA   
19                health  operating  2007-10-01 2013-08-13          USA   
20             cleantech  operating  2003-01-01 2012-01-26          USA   
...                  ...        ...         ...        ...          ...   
196537            social  operating  2011-01-11 2011-09-06          CAN   
196540       games_video     closed  2006-06-01 2009-02-01          USA   
196547       advertising  operating  2007-01-01 2008-05-01          NaN   
196549  public_relations  operating  2007-11-01 2008-08-25          USA   
196552         ecommerce  operating  2007-09-01 2008-06-01          USA   

       state_code               city             region  investment_rounds  \
0              WA            Seattle            Seattle                  0   
13             CA      Mountain View             SF Bay                  0   
18             CA          Palo Alto             SF Bay                  0   
19             CA      San Francisco             SF Bay                  0   
20             TX             Austin             Austin                  0   
...           ...                ...                ...                ...   
196537        NaN           Victoria           Victoria                  0   
196540         CA      San Francisco             SF Bay                  0   
196547        NaN                NaN            unknown                  0   
196549         CA           San Jose             SF Bay                  0   
196552         CA  Pacific Palisades  Pacific Palisades                  0   

        invested_companies  funding_rounds  funding_total_usd  \
0                        0               3         39750000.0   
13                       0               1          5000000.0   
18                       0               1                0.0   
19                       0               5         68069200.0   
20                       0               3         10125293.0   
...                    ...             ...                ...   
196537                   0               1           500000.0   
196540                   0               2          1100000.0   
196547                   0               2           500000.0   
196549                   0               1           750000.0   
196552                   0               1           475000.0   

       first_milestone_at last_milestone_at  milestones  relationships  
0              2010-09-05        2013-09-18           5             17  
13             2008-05-01        2012-09-13           3             14  
18             1995-03-01        2012-09-18           4              9  
19                    NaT               NaT           0             14  
20             2010-01-01        2010-01-01           1              6  
...                   ...               ...         ...            ...  


In [104]:
# Checking funding_rounds_clean.csv for more end dates

fr_df = pd.read_csv(path + 'funding_rounds_clean.csv')
fr_df.head()

Unnamed: 0  funding_round_id object_id  funded_at funding_round_type  \
0           0                 1       c:4  12/1/2006           series-b   
1           1                 2       c:5   9/1/2004              angel   
2           2                 3       c:5   5/1/2005           series-a   
3           3                 4       c:5   4/1/2006           series-b   
4           4                 5    c:7299   5/1/2006           series-b   

  funding_round_code  raised_amount_usd  raised_amount raised_currency_code  \
0                  b            8500000        8500000                  USD   
1              angel             500000         500000                  USD   
2                  a           12700000       12700000                  USD   
3                  b           27500000       27500000                  USD   
4                  b           10500000       10500000                  USD   

   pre_money_valuation_usd  pre_money_valuation pre_money_currency_code  \
0                        0                    0                     NaN   
1                        0                    0                     USD   
2                115000000            115000000                     USD   
3                525000000            525000000                     USD   
4                        0                    0                     NaN   

   post_money_valuation_usd  post_money_valuation post_money_currency_code  \
0                         0                     0                      NaN   
1                         0                     0                      USD   
2                         0                     0                      USD   
3                         0                     0                      USD   
4                         0                     0                      NaN   

   participants  is_first_round  is_last_round  
0             2               0              0  
1             2               0              1  
2             3               0              0  
3             4               0              0  
4             2               0              0

In [105]:
fr_df.columns.to_list()

['Unnamed: 0',
 'funding_round_id',
 'object_id',
 'funded_at',
 'funding_round_type',
 'funding_round_code',
 'raised_amount_usd',
 'raised_amount',
 'raised_currency_code',
 'pre_money_valuation_usd',
 'pre_money_valuation',
 'pre_money_currency_code',
 'post_money_valuation_usd',
 'post_money_valuation',
 'post_money_currency_code',
 'participants',
 'is_first_round',
 'is_last_round']

##### There doesn't seem to be any column that would provide us with end dates 
I would need to compare the funded_at column with the start dates from Irina's notebook to see if there is any overlap. I'm thinking that the funded_at cloumns will have a later date than the start dates.

In [106]:
# Adding new column for collecting start dates
fr_df["start_dates"] = ""
fr_df.head()

Unnamed: 0  funding_round_id object_id  funded_at funding_round_type  \
0           0                 1       c:4  12/1/2006           series-b   
1           1                 2       c:5   9/1/2004              angel   
2           2                 3       c:5   5/1/2005           series-a   
3           3                 4       c:5   4/1/2006           series-b   
4           4                 5    c:7299   5/1/2006           series-b   

  funding_round_code  raised_amount_usd  raised_amount raised_currency_code  \
0                  b            8500000        8500000                  USD   
1              angel             500000         500000                  USD   
2                  a           12700000       12700000                  USD   
3                  b           27500000       27500000                  USD   
4                  b           10500000       10500000                  USD   

   pre_money_valuation_usd  pre_money_valuation pre_money_currency_code  \
0                        0                    0                     NaN   
1                        0                    0                     USD   
2                115000000            115000000                     USD   
3                525000000            525000000                     USD   
4                        0                    0                     NaN   

   post_money_valuation_usd  post_money_valuation post_money_currency_code  \
0                         0                     0                      NaN   
1                         0                     0                      USD   
2                         0                     0                      USD   
3                         0                     0                      USD   
4                         0                     0                      NaN   

   participants  is_first_round  is_last_round start_dates  
0             2               0              0              
1             2               0              1              
2             3               0              0              
3             4               0              0              
4             2               0              0

In [107]:
fr_df.columns.to_list()

['Unnamed: 0',
 'funding_round_id',
 'object_id',
 'funded_at',
 'funding_round_type',
 'funding_round_code',
 'raised_amount_usd',
 'raised_amount',
 'raised_currency_code',
 'pre_money_valuation_usd',
 'pre_money_valuation',
 'pre_money_currency_code',
 'post_money_valuation_usd',
 'post_money_valuation',
 'post_money_currency_code',
 'participants',
 'is_first_round',
 'is_last_round',
 'start_dates']

In [108]:
fr_df = fr_df[['funding_round_id',
 'object_id',
 'start_dates',              
 'funded_at',
 'funding_round_type',
 'funding_round_code',
 'raised_amount_usd',
 'raised_amount',
 'raised_currency_code',
 'pre_money_valuation_usd',
 'pre_money_valuation',
 'pre_money_currency_code',
 'post_money_valuation_usd',
 'post_money_valuation',
 'post_money_currency_code',
 'participants',
 'is_first_round',
 'is_last_round',]]

In [109]:
fr_df['start_dates'] = fr_df.apply(lambda row : row['funded_at'], axis = 1)
fr_df[fr_df['start_dates'].notna()]
fr_df

funding_round_id object_id start_dates   funded_at funding_round_type  \
0                     1       c:4   12/1/2006   12/1/2006           series-b   
1                     2       c:5    9/1/2004    9/1/2004              angel   
2                     3       c:5    5/1/2005    5/1/2005           series-a   
3                     4       c:5    4/1/2006    4/1/2006           series-b   
4                     5    c:7299    5/1/2006    5/1/2006           series-b   
...                 ...       ...         ...         ...                ...   
52923             57948  c:211890  12/12/2013  12/12/2013           series-a   
52924             57949  c:267427    2/6/2010    2/6/2010            venture   
52925             57950  c:261728    2/6/2010    2/6/2010            venture   
52926             57951  c:285864  12/12/2013  12/12/2013           series-a   
52927             57952  c:286215    4/7/2010    4/7/2010            venture   

      funding_round_code  raised_amount_usd  raised_amount  \
0                      b            8500000        8500000   
1                  angel             500000         500000   
2                      a           12700000       12700000   
3                      b           27500000       27500000   
4                      b           10500000       10500000   
...                  ...                ...            ...   
52923                  a            3000000        3000000   
52924            partial             570000         570000   
52925       unattributed            2184100        2184100   
52926                  a             790783         790783   
52927            partial             271250         271250   

      raised_currency_code  pre_money_valuation_usd  pre_money_valuation  \
0                      USD                        0                    0   
1                      USD                        0                    0   
2                      USD                115000000            115000000   
3                      USD                525000000            525000000   
4                      USD                        0                    0   
...                    ...                      ...                  ...   
52923                  USD                        0                    0   
52924                  USD                        0                    0   
52925                  USD                        0                    0   
52926                  USD                        0                    0   
52927                  USD                        0                    0   

      pre_money_currency_code  post_money_valuation_usd  post_money_valuation  \
0                         NaN                         0                     0   
1                         USD                         0                     0   
2                         USD                         0                     0   
3                         USD                         0                     0   
4                         NaN                         0                     0   
...                       ...                       ...                   ...   
52923                     USD                         0                     0   
52924                     USD                         0                     0   
52925                     USD                         0                     0   
52926                     USD                         0                     0   
52927                     USD                         0                     0   

      post_money_currency_code  participants  is_first_round  is_last_round  
0                          NaN             2               0              0  
1                          USD             2               0              1  
2                          USD             3               0              0  
3                          USD             4               0              0  
4                        

what columns should I drop? 
Note to self, need to make a func to get min date